week2/assignment2_3_2/src/sensor_data.cpp

```diff
@@ -0,0 +1,187 @@
+{
+    return std::abs(a - b);
+}
+//Function to get standard deviation for sensor_values
+float get_value_std_deviation(const std::vector<Sensor_reading>& data)
+{
+    float sum_of_distances_squared{ 0.0f };
+    float mean{ get_value_mean(data) };
+    for (const auto& reading : data)
+    {
+        sum_of_distances_squared += std::pow(get_absolute_difference(reading.m_sensor_value, mean), 2.0);
+    }
+    return std::sqrt(sum_of_distances_squared / data.size());
+}
+//Function checks if value is outlier
+bool value_is_outlier(float distance, float std_deviation)
+{
+    return distance > std_deviation;
+}
+
+//A function for checking outliers
+//Checks and sets the flags for m_is_outlier and m_outlier_checked accordingly
+void check_for_outliers(std::vector<Sensor_reading>&data)
+{
+    float mean{ get_value_mean(data) };
+    float std_deviation{ get_value_std_deviation(data) };
+    int counter{ 0 };
+    for (auto& reading : data)
+    {
+        if (get_absolute_difference(mean, reading.m_sensor_value) > std_deviation)
+        {
+            reading.m_is_outlier = true;
+            std::cout << "Outlier found: " << reading << '\n';
+            ++counter;
+        }
+        reading.m_outlier_checked = true;
+    }
+    std::cout << "Outliers found: " << counter << '\n';
+}
+//A function that checks which sensor(s) are the most active, giving out most
+//sensor readings - returns a vector of strings with sensor_ids
+std::vector<std::string> most_frequent_sensor(std::vector<Sensor_reading>& data)
+{
+    //Sort the data according to sensor_id
+    std::sort(data.begin(), data.end(), [](const auto& a, const auto& b)
+        {
+            return a.m_sensor_id < b.m_sensor_id;
+        });
+    
+    int max_count{ 0 };
+    std::vector<std::string> max_values{};
+    int current_count{ 0 };
+
+    //Loop through the data once,  as we have ensured its sorted
+    for (auto it{ data.begin() }; it != data.end();++it )
+    {
+        //Value to test against
+        std::string current_value{it->m_sensor_id};
+        
+        //Fast forward and count as long as value stays the same or next is not .end()
+        while ((it + 1)->m_sensor_id == current_value && (it + 1) != data.end())
+        {
+            ++current_count;
+            ++it;
+
+        }
+        //Add the last count after while loop 
+        ++current_count;
+        
+        //If the new count is as big as max_count, add to vector
+        if (current_count == max_count)
+        {
+            max_values.push_back(current_value);
+        }
+        //If the new count is bigger than max_count, clear and add to vector
+        else if (current_count > max_count)
+        {
+            max_values.clear();
+            max_values.push_back(current_value);
+            max_count = current_count;
+        }
+        //Reset current_count when going to next value
+        current_count = 0;
+    }
+    return max_values;
+}
```

Comment:
You could also use:
// find first not matching spot
auto uEnd = std::find_if_not(it, end);
// calculate distance from current it value to non matching spot (=count)
count = std::distance(it, uEnd);
// advance by a count
std::advance(it, count)